### PR TITLE
State manager

### PR DESCRIPTION
**Pull request by Juliusan**
Refactored state manager synchronisation algorithm. As block essence hash is included in alias output, the catching up algorithm may be changed to deterministically obtain block sequence from the last block to the first one. There is no need to try choosing every possible block candidate from the first one to the last one and check if the obtained sequence is correct.

Unit tests in `stardust-test` script passes as well as deploy cluster tests. `TestMissingRequests` was passing until I merged recent stardust-vm changes.

### CODE REVIEW

**Review by lunfardo314 — APPROVED** [review]

File: packages/chain/statemgr/action.go [hunk 1]
```diff
@@ -109,7 +109,9 @@ func (sm *stateManager) addStateCandidateFromConsensus(nextState state.VirtualSt
 		return false
 	}
 	block.SetApprovingOutputID(approvingOutputID)
-	sm.addBlockAndCheckStateOutput(block, nextState)
+	sm.syncingBlocks.startSyncingIfNeeded(nextState.BlockIndex())
+	sm.syncingBlocks.addBlockCandidate(block, nextState)
```
Comment (jorgemmsilva) [inline packages/chain/statemgr/action.go:113]: I don't fully grok how this process works, but shouldn't we `addBlockCandidate` before `startSyncingIfNeeded`?

File: packages/util/permute_test.go [hunk 2]
```diff
@@ -4,16 +4,93 @@ import (
...
+	// There is p = 1 - 3628800!/(3628795!*3628800^5) ~ 0.00028% probability of obtaining at least two equal permutations while picking 5 random ones.
+	// There is q = 1 - (1-p)^10 ~ 0.0028% probability of obtaining at least two equal permutations while picking 5 random ones if you have 10 tries to do that.
+	// q is a probability of test returning false positive on initial permutation length of 10. ~ 1 false positive in 36k runs.
+	// For 9 q ~ 0.028% ~ 1 false positive in 3.6k runs.
+	// For 8 q ~ 0.25% ~ 1 false positive in 400 runs.
+	// For 7 q ~ 2.0% ~ 1 false positive in 50 runs.
+	// For 6 q ~ 13.0% ~ 1 false positive in 8 runs.
+	// For 5 q ~ 57.0% ~  false positive every second run.
+	// That is why `n` starts at 10.
+	// NOTE: the exact values might be a little bit of due to accuracy of float arythmetics in LibreOffice Calc. They are provided just to get an impression.
```
Comment (jorgemmsilva) [inline packages/util/permute_test.go:66]: do we just accept that we have a test  with a small probability of failing? :/

File: stardust-test [hunk 3]
```diff
@@ -26,6 +26,8 @@ PATHS=(\
     "packages/dkg/..." \
     "packages/iscp" \
     "packages/webapi/..." \
+    "packages/util" \
+    "packages/util/pipe" \
```
Comment (jorgemmsilva) [inline stardust-test]:
```suggestion
    "packages/util/..." \
```
would this work?

**Review by jorgemmsilva — COMMENTED** [review]
it looks fine to me overall, but the logic around the sync algorigthm is a bit hard to follow. Would you mind adding a small .md file with an high level description of how it should work? :pray:

**Review by jorgemmsilva — APPROVED** [review]

File: stardust-test (hunk 3 above)
Comment (Juliusan) [reply stardust-test]: Not sure about all the packages. Will check.

File: packages/util/permute_test.go (hunk 2 above)
Comment (Juliusan) [reply packages/util/permute_test.go:66]: IMO you cannot do better. Either you remove the test or you accept that it might theoretically fail. If you try to test that each time a random algorithm picks two different values (or permutations in this case), you have to accept that sometimes randomly same values are picked consecutively. Or you have to improve the random algorithm to ignore the previously picked value.

File: packages/chain/statemgr/action.go (hunk 1 above)
Comment (Juliusan) [reply packages/chain/statemgr/action.go:113]: I'll do the MD file. No. `startSyncingIfNeeded` simply creates structures to store blocks in. To avoid memory leaks these structures are later destroyed. `addBlockCandidate` adds a block to the structure, which must be created in advance.
